pkg/oidc/oidc.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oidc
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+const (
+	defaultClientID      = "gangway"
+	defaultUsernameClaim = "email"
+	defaultGroupsClaim   = "groups"
+)
+
+// Config deals with providing OIDC related fields to the Lokomotive
+// cluster for OIDC based Authentication via Dex and Gangway.
+type Config struct {
+	IssuerURL     string `hcl:"issuer_url,optional"`
+	ClientID      string `hcl:"client_id,optional"`
+	UsernameClaim string `hcl:"username_claim,optional"`
+	GroupsClaim   string `hcl:"groups_claim,optional"`
+}
+
+// ConfigureOIDCFlags configures the Config fields after setting default values
+// and returning a list of oidc flags and error if any.
+func (c *Config) ConfigureOIDCFlags(clusterDomain string) []string {
+	// Set defaults if values not provided in the configuration.
+	if c.ClientID == "" {
+		c.ClientID = defaultClientID
+	}
+
+	if c.UsernameClaim == "" {
+		c.UsernameClaim = defaultUsernameClaim
+	}
+
+	if c.GroupsClaim == "" {
+		c.GroupsClaim = defaultGroupsClaim
+	}
+
+	if c.IssuerURL == "" {
+		c.IssuerURL = fmt.Sprintf("https://dex.%s", clusterDomain)
+	}
+
+	// initialize map
+	oidcFlags := map[string]string{
+		"--oidc-issuer-url":     c.IssuerURL,
+		"--oidc-client-id":      c.ClientID,
+		"--oidc-username-claim": c.UsernameClaim,
+		"--oidc-groups-claim":   c.GroupsClaim,
+	}
+
+	oidcFlagsList := []string{}
+	// Convert key,value pair to an item in a slice.
+	for k, v := range oidcFlags {
+		oidcFlagsList = append(oidcFlagsList, fmt.Sprintf("\"%s=%s\"", k, v))
```

Comment:
I suggest we avoid nesting quotes in string if possible, as it might be confusing when reading code. Instead, we add quotes while rendering the list. This also removes the assumption, that consumer of output requires already quoted strings.

```suggestion
		oidcFlagsList = append(oidcFlagsList, fmt.Sprintf("%s=%s", k, v))
```